integration/dkg/dkg_test.go

```diff
@@ -0,0 +1,248 @@
+			core.Log,
+			core.Me,
+			NewWhiteboardClient(id.NodeID, whiteboard),
+			brokerTunnel,
+		),
+		viewsObserver,
+	)
+
+	node := node{
+		GenericNode:     core,
+		keyStorage:      dkgKeys,
+		messagingEngine: messagingEngine,
+		reactorEngine:   reactorEngine,
+	}
+
+	return &node
+}
+
+func TestDKG(t *testing.T) {
+
+	// hub is an in-memory test network that enables nodes to communicate using
+	// the DKG messaging engine
+	hub := stub.NewNetworkHub()
+
+	// whiteboard is a shared object where DKG nodes can publish/read broadcast
+	// messages, as well as publish end results, using a special
+	// DKGContractClient.
+	// TODO: replace with a real smart-contract and emulator
+	whiteboard := newWhiteboard()
+
+	chainID := flow.Testnet
+
+	// we run the DKG protocol with 10 consensus nodes
+	conIdentities := unittest.IdentityListFixture(10, unittest.WithRole(flow.RoleConsensus))
+
+	// The EpochSetup event is received at view 100. The phase transitions are
+	// at views 150, 200, and 250. In between phase transitions, the controller
+	// calls the DKG smart-contract every 10 views.
+	//
+	// VIEWS
+	// setup      : 100
+	// polling    : 110 120 130 140 150
+	// Phase1Final: 150
+	// polling    : 160 170 180 190 200
+	// Phase2Final: 200
+	// polling    : 210 220 230 240 250
+	// Phase3Final: 250
+	// final      : 300
+
+	// create and record relevant blocks
+	blocks := make(map[uint64]*flow.Header)
+	var view uint64
+	for view = 100; view <= 250; view += dkgeng.DefaultPollStep {
+		blocks[view] = &flow.Header{View: view}
+	}
+	firstBlock := blocks[100]
+
+	// create the EpochSetup that will trigger the next DKG run with all the
+	// desired parameters
+	epochSetup := flow.EpochSetup{
+		Counter:            999,
+		DKGPhase1FinalView: 150,
+		DKGPhase2FinalView: 200,
+		DKGPhase3FinalView: 250,
+		FinalView:          300,
+		Participants:       conIdentities,
+		RandomSource:       []byte("random bytes for seed"),
+	}
+
+	nodes, _ := createNodes(
+		t,
+		hub,
+		chainID,
+		whiteboard,
+		conIdentities,
+		epochSetup,
+		firstBlock.ID())
+
+	for _, n := range nodes {
+		n.Ready()
+	}
+
+	// trigger the EpochSetupPhaseStarted event for all nodes, effectively
+	// starting the next DKG run
+	for _, n := range nodes {
+		n.reactorEngine.EpochSetupPhaseStarted(epochSetup.Counter, firstBlock)
+	}
+
+	// trigger the BlockFinalized events for each view of interest, effectively
+	// causing the DKG state machine to progress
+	for view = 100; view <= 250; view += dkgeng.DefaultPollStep {
+		time.Sleep(100 * time.Millisecond)
+		hub.DeliverAll()
+		for _, n := range nodes {
+			n.ProtocolEvents.BlockFinalized(blocks[view])
+		}
+	}
+
+	for _, n := range nodes {
+		n.Done()
+	}
+
+	// check that all nodes have published the same DKG result
+	require.Equal(t, 1, len(whiteboard.resultSubmitters))
+	for _, signers := range whiteboard.resultSubmitters {
+		require.Equal(t, 10, len(signers))
+	}
+
+	// check that each node has saved a private key
```

Comment:
We should also attempt to sign and verify using the keys, using the stored private keys and the public key we computed